hazelcast/src/main/java/com/hazelcast/spi/impl/operationexecutor/slowoperationdetector/SlowOperationDetector.java

```diff
@@ -183,45 +184,57 @@ private void scanOperationRunner(long nowNanos, long nowMillis, OperationRunner
                 return;
             }
 
-            // create the stack trace once for this operation and cache the invocation for upcoming updates
-            SlowOperationLog log = getOrCreateSlowOperationLog(operationRunner);
-            operationData.invocation = log.getOrCreateInvocation(operationHashCode, durationNanos, nowNanos, nowMillis);
+            // create the stack trace once (if the running operation didn't change)
+            String stackTrace = getStackTraceOrNull(operationRunner, operation);
+            if (stackTrace != null) {
+                // create the log for this operation and cache the invocation for upcoming updates
+                SlowOperationLog log = getOrCreateSlowOperationLog(stackTrace, operation);
+                int totalInvocations = log.totalInvocations.incrementAndGet();
+                operationData.invocation = log.getOrCreateInvocation(operationHashCode, durationNanos, nowNanos, nowMillis);
 
-            logSlowOperation(operationRunner, log);
+                logSlowOperation(operation, log, totalInvocations);
+            }
         }
 
-        private SlowOperationLog getOrCreateSlowOperationLog(final OperationRunner operationRunner) {
+        private String getStackTraceOrNull(OperationRunner operationRunner, Object operation) {
             String prefix = "";
             for (StackTraceElement stackTraceElement : operationRunner.currentThread().getStackTrace()) {
                 stackTraceStringBuilder.append(prefix).append(stackTraceElement.toString());
                 prefix = "\n\t";
             }
 
-            String stackTraceString = stackTraceStringBuilder.toString();
+            // check if the operation is still the same
+            if (operationRunner.currentTask() != operation) {
+                stackTraceStringBuilder.setLength(0);
+                return null;
+            }
+
+            String stackTrace = stackTraceStringBuilder.toString();
             stackTraceStringBuilder.setLength(0);
+            return stackTrace;
+        }
 
-            Integer hashCode = stackTraceString.hashCode();
-            SlowOperationLog candidate = slowOperationLogs.get(hashCode);
+        private SlowOperationLog getOrCreateSlowOperationLog(String stackTrace, Object operation) {
```

Comment:
What I often do is to base the naming on the context. Since you are doing stuff with slow operation logs and you know the return type, I often go for a shorter name like 'getOrCreate'. 

This prevents running into those long lines you need to chew through.